lib/std/math/atan2.zig

```diff
@@ -10,25 +10,28 @@ const expect = std.testing.expect;
 
 /// Returns the arc-tangent of y/x.
 ///
-/// Special Cases:
-///  - atan2(y, nan)     = nan
-///  - atan2(nan, x)     = nan
-///  - atan2(+0, x>=0)   = +0
-///  - atan2(-0, x>=0)   = -0
-///  - atan2(+0, x<=-0)  = +pi
-///  - atan2(-0, x<=-0)  = -pi
-///  - atan2(y>0, 0)     = +pi/2
-///  - atan2(y<0, 0)     = -pi/2
-///  - atan2(+inf, +inf) = +pi/4
-///  - atan2(-inf, +inf) = -pi/4
-///  - atan2(+inf, -inf) = 3pi/4
-///  - atan2(-inf, -inf) = -3pi/4
-///  - atan2(y, +inf)    = 0
-///  - atan2(y>0, -inf)  = +pi
-///  - atan2(y<0, -inf)  = -pi
-///  - atan2(+inf, x)    = +pi/2
-///  - atan2(-inf, x)    = -pi/2
-pub fn atan2(comptime T: type, y: T, x: T) T {
+///      Special Cases:
+/// |   y   |   x   | radians |
+/// |-------|-------|---------|
+/// |  fin  |  nan  |   nan   |
+/// |  nan  |  fin  |   nan   |
+/// |  +0   | >=0   |   +0    |
+/// |  -0   | >=0   |   -0    |
+/// |  +0   |<=-0   |   pi    |
+/// |  -0   |<=-0   |  -pi    |
+/// |  >0   |   0   |  +pi/2  |
+/// |  <0   |   0   |  -pi/2  |
+/// | +inf  | +inf  |  +pi/4  |
+/// | -inf  | +inf  |  -pi/4  |
+/// | +inf  | -inf  |  3pi/4  |
+/// | -inf  | -inf  | -3pi/4  |
+/// |  fin  | +inf  |    0    |
+/// |  >0   | -inf  |  +pi    |
+/// |  <0   | -inf  |  -pi    |
+/// | +inf  |  fin  |  +pi/2  |
+/// | -inf  |  fin  |  -pi/2  |
+pub fn atan2(y: anytype, x: @TypeOf(y)) @TypeOf(y) {
+    const T = @TypeOf(y);
```

Comment:
The current signature will result in a "unable to resolve comptime value" if the user calls it with a `comptime_float` or literal as the first argument but a runtime float value as the second argument, which will make it a bit awkward to use in those cases.

```zig
const value: f32 = ...;

// this is fine
const result = atan2(value, 3);

// error: unable to resolve comptime value
const result = atan2(3, value);

// the literal must be wrapped in '@as'
const result = atan2(@as(f32, 3), value);
```

It might be a better idea to use multi-argument `@TypeOf` to invoke peer type resolution, since it will automatically handle the coercion from `comptime_float` to runtime floats.

```suggestion
pub fn atan2(y: anytype, x: anytype) @TypeOf(y, x) {
    const T = @TypeOf(y, x);
```

Reply:
Yes, that makes sense to me. `std.math.Complex` guarantees that both fields have the same type but since the standalone `std.math.complex.arg` function is duck typed it makes sense to check both.

Reply:
Just FYI, I'm just some rando offering suggestions, not a core member of the core Zig team, so it's ultimately up to them to review and decide what they want to do with this PR.

Personally, as of the most recent commit I think the PR might be trying to do too many unrelated things at once. You might have a better chance of getting this reviewed and merged if you if you split updating `atan2` (+ `arg` since there's an argument to be made that it's closely related to atan2) and updating the remainder of the `complex` functions into two separate PR, so that any eventual concerns with one aspect won't block the other.

(You removed your comment and made major changes to the PR while I was typing but you made the same call as I was suggesting so I guess we both agree 😄)